d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/RelativeStateUpdater.java

```diff
@@ -0,0 +1,427 @@
+  {
+    _relativeStrategyProperties = relativeStrategyProperties;
+    _quarantineManager = quarantineManager;
+    _executorService = executorService;
+    _listenerFactories = listenerFactories;
+    _partitionLoadBalancerStateMap = partitionLoadBalancerStateMap;
+    _lock = new ReentrantLock();
+
+    _executorService.scheduleWithFixedDelay((Runnable) this::updateState, EXECUTOR_INITIAL_DELAY,
+        _relativeStrategyProperties.getUpdateIntervalMs(), TimeUnit.MILLISECONDS);
+  }
+
+  @Override
+  public void updateState(Set<TrackerClient> trackerClients, int partitionId, long clusterGenerationId)
+  {
+    if (!_partitionLoadBalancerStateMap.containsKey(partitionId))
+    {
+      // If the partition is not initialized, initialize the state synchronously
+      initializePartition(trackerClients, partitionId, clusterGenerationId);
+    }
+    else if (clusterGenerationId != _partitionLoadBalancerStateMap.get(partitionId).getClusterGenerationId())
+    {
+      // Asynchronously update the state if it is from uri properties change
+      PartitionState oldPartitionState = _partitionLoadBalancerStateMap.get(partitionId);
+      _executorService.execute(() -> updateStateForPartition(trackerClients, partitionId, oldPartitionState, clusterGenerationId));
+    }
+  }
+
+  @Override
+  public Ring<URI> getRing(int partitionId)
+  {
+    return _partitionLoadBalancerStateMap.get(partitionId) == null
+        ? null
+        : _partitionLoadBalancerStateMap.get(partitionId).getRing();
+  }
+
+  Map<URI, Integer> getPointsMap(int partitionId)
+  {
+    return _partitionLoadBalancerStateMap.get(partitionId) == null
+        ? new HashMap<>()
+        : _partitionLoadBalancerStateMap.get(partitionId).getPointsMap();
+  }
+
+  /**
+   * Update the partition state.
+   * This is scheduled by executor, and the update should not involve any cluster change
+   */
+  void updateState()
+  {
+    // Update state for each partition
+    for (Integer partitionId : _partitionLoadBalancerStateMap.keySet())
+    {
+      PartitionState partitionState = _partitionLoadBalancerStateMap.get(partitionId);
+      if (partitionState != null)
+      {
+        updateStateForPartition(partitionState.getTrackerClients(), partitionId, partitionState, null);
+      }
+    }
+  }
+
+  /**
+   * Update the partition state when there is a cluster uris change.
+   */
+  void updateStateForPartition(Set<TrackerClient> trackerClients, int partitionId,
+      PartitionState oldPartitionState, Long clusterGenerationId)
+  {
+    LOG.debug("Updating for partition: " + partitionId + ", state: " + oldPartitionState);
+    PartitionState newPartitionState = oldPartitionState.copy();
+
+    // Step 1: Update the base health scores for each {@link TrackerClient} in the cluster
+    Map<TrackerClient, CallTracker.CallStats> latestCallStatsMap = new HashMap<>();
+    long avgClusterLatency = getAvgClusterLatency(trackerClients, latestCallStatsMap);
```

Comment:
I exacted avgClusterLatency because it is also used in quarantine calculation later, if I put it into #updateHealthScoreAndState, it's hard for me to pass the value into quarantine method.